src/java.base/share/classes/jdk/internal/foreign/SegmentFactories.java

```diff
@@ -204,7 +204,7 @@ private static long allocateNativeInternal(long byteSize, long byteAlignment, Me
         long allocationBase;
         long result;
         if (byteAlignment > MAX_MALLOC_ALIGN || alignedSize % byteAlignment != 0) {
-            allocationSize = alignedSize + byteAlignment - MAX_MALLOC_ALIGN;
+            allocationSize = alignedSize + byteAlignment - Math.min(MAX_MALLOC_ALIGN, alignedSize);
```

Comment:
This doesn't look correct to me. Let's say we want to do a 7 byte allocation (with `init == false`), aligned to 4. We get `7 % 4 -> 3` so we enter this branch, and then `allocationSize = 7 + 4 - 7`, and we end up allocating only 4 bytes. The previous computation only works if `MAX_MALLOC_ALIGN` is greater than `byteAlignment`.

I think it's simpler to just replace `MAX_MALLOC_ALIGN` in this `if` statement with an 'expected alignment' derived from the size.